src/timeout.ts

```diff
@@ -0,0 +1,100 @@
+import { clearTimeout, setTimeout } from 'timers';
+
+import { MongoInvalidArgumentError } from './error';
+import { noop } from './utils';
+
+/** @internal */
+export class TimeoutError extends Error {
+  override get name(): 'TimeoutError' {
+    return 'TimeoutError';
+  }
+
+  constructor(message: string, options?: { cause?: Error }) {
+    super(message, options);
+  }
+
+  static is(error: unknown): error is TimeoutError {
+    return (
+      error != null && typeof error === 'object' && 'name' in error && error.name === 'TimeoutError'
+    );
+  }
+}
+
+type Executor = ConstructorParameters<typeof Promise<never>>[0];
+type Reject = Parameters<ConstructorParameters<typeof Promise<never>>[0]>[1];
+/**
+ * @internal
+ * This class is an abstraction over timeouts
+ * The Timeout class can only be in the pending or rejected states. It is guaranteed not to resolve
+ * if interacted with exclusively through its public API
+ * */
+export class Timeout extends Promise<never> {
+  get [Symbol.toStringTag](): 'MongoDBTimeout' {
+    return 'MongoDBTimeout';
+  }
+
+  private timeoutError: TimeoutError;
+  private id?: NodeJS.Timeout;
+
+  public readonly start: number;
+  public ended: number | null = null;
+  public duration: number;
+  public timedOut = false;
+
+  /** Create a new timeout that expires in `duration` ms */
+  private constructor(executor: Executor = () => null, duration: number) {
+    let reject!: Reject;
+
+    if (duration < 0) {
+      throw new MongoInvalidArgumentError('Cannot create a Timeout with a negative duration');
+    }
+
+    super((_, promiseReject) => {
+      reject = promiseReject;
+
+      executor(noop, promiseReject);
+    });
+
+    // NOTE: Construct timeout error at point of Timeout instantiation to preserve stack traces
+    this.timeoutError = new TimeoutError(`Expired after ${duration}ms`);
```

Comment:
How does this work with CSOT, when a single timeout will be reused for multiple operations?

Reply:
looked more closely at this class - if this is internal, I don't see why capturing it in the constructor for the stack trace is important

Reply:
In keeping with my response to your other question, I'm open to just throwing a string.

Reply:
Also as far as the message with the duration string is concerned, part of the use case for CSOT would be, for each of the blocking sections, we construct a new Timeout from the Timeout constructed for the whole operation that either uses the remaining timeoutMS or another timeout value as detailed in the spec. For example, when doing server selection, the [spec](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#server-selection) requires that we timeout that operation with the remaining timeoutMS or serverSelectionTimeoutMS. Which would look something like

```ts
const timeout = options.timeout
					? Timeout.expires(min(options.timeout.remainingTime, options.serverSelectionTimeoutMS))
					: Timeout.expires(options.serverSelectionTimeoutMS)
```

Reply:
Okay, that works.  Early CSOT designs re-used timeouts (or timeout controllers) and just reset the internal timeout instead of creating a new timeout each time, which would make the stack trace incorrect.  If we're creating a new timeout for every async part of the driver, that's not an issue.